File: solutions/419/main.py
from typing import List 

class Solution:
    def dfs(self, board, i, j):
        # base case
        if i >= len(board) or i < 0:
            return
        if j >= len(board[0]) or j < 0:
            return
        if board[i][j] == ".":
            return
        board[i][j] = "."
        # traverse
        self.dfs(board, i+1, j)
        self.dfs(board, i-1, j)
        self.dfs(board, i, j+1)
        self.dfs(board, i, j-1)
    def countBattleships(self, board: List[List[str]]) -> int:
        result = 0
        for i in range(len(board)):
            for j in range(len(board[0])):
                if board[i][j] == "X":
                    result += 1
                    self.dfs(board, i, j)
        return result

if __name__ == "__main__": 
    board = [["X",".",".","X"],[".",".",".","X"],[".",".",".","X"]]
    s = Solution()
    assert s.countBattleships(board) == 2 

File: solutions/419/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Fri Dec 17 14:12:23 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
public:
    int countBattleships(vector<vector<char>>& board) {
        int result = 0;
        for (int i = 0; i < board.size(); ++i) {
            for (int j = 0; j < board[0].size(); ++j) {
                if (board[i][j] == 'X') {
                    if (i-1 < 0 || (i-1 >= 0 && board[i-1][j] == '.')) {
                        if (j-1 < 0 || j-1 >= 0 && board[i][j-1] == '.')
                            result++;
                    }
                }
            }
        }
        return result;
    }
};

TEST(Test419, SimpleTest) {
    vector<vector<char>> board{{'X','.','.','X'},{'.','.','.','X'},{'.','.','.','X'}};
    Solution s; 
    EXPECT_EQ(s.countBattleships(board), 2);
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}
